### PR TITLE
[Return-types #6] Update adjoint differentiation

### PR DESCRIPTION
**Pull request by eddddddy**
**Context:**
Part of the ongoing effort to support gradient methods with the new return types system. This PR focuses on supporting the adjoint differentiation method.

**Description of the Change:**
Add a new method `adjoint_jacobian_new` to `QubitDevice` that computes the gradient using the adjoint method, and returns the result as a tuple.

**Example:**

```python
dev = qml.device("default.qubit", wires=2)

with qml.tape.QuantumTape() as tape:
    qml.RX(0.5, wires=0)
    qml.CNOT(wires=[0, 1])
    qml.RY(0.8, wires=1)
    qml.expval(qml.PauliZ(0))
    qml.expval(qml.PauliZ(1))
    qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
```
```pycon
>>> dev.adjoint_jacobian_new(tape)
(array([-4.79425539e-01, -3.46944695e-18]), array([-0.33401899, -0.6295392 ]), array([ 0.        , -0.71735609]))
```

### CODE REVIEW

File: pennylane/_qubit_device.py
Comment (antalszava) [inline pennylane/_qubit_device.py:1891]: Let's temporarily make a comment that this is the version of `adjoint_jacobian` for the new return type system.

File: pennylane/_qubit_device.py
Comment (antalszava) [inline pennylane/_qubit_device.py:2023]: Are these lines the only diff from `adjoint_jacobian`? If so, `adjoint_jacobian_new` could call `adjoint_jacobian` and post-process its result to ensure an ease of maintainability. This would be a fortunate case until no other part of the logic requires change.

File: tests/returntypes/test_adjoint_diff_new.py
Comment (antalszava) [inline tests/returntypes/test_adjoint_diff_new.py:189]: Let's check that `dev_jacobian` is a `tuple` and that each element within is an array of expected shapes. This will be relevant for other test cases too.

File: tests/returntypes/test_adjoint_diff_new.py
Comment (antalszava) [inline tests/returntypes/test_adjoint_diff_new.py:365]: Minor: we may as well check the type and shape of every tuple element as extra checks.
```suggestion

        # check the results against individually executed tapes
        for i, ob in enumerate(observables):
            with qml.tape.QuantumTape() as indiv_tape:
                for op in ops:
                    qml.apply(op)

                qml.expval(ob)

            indiv_tape.trainable_params = {1, 2, 3}

            assert isinstance(grad_D[i], np.ndarray)
            assert grad_D[i].shape == (3,)
            assert np.allclose(grad_D[i], dev.adjoint_jacobian_new(indiv_tape))
```

**Review by antalszava — COMMENTED** [review]
This is looking great already! :tada: I think it may just be worth considering how best to include the logic here (potentially by only duplicating code that definitely needs repetition), but otherwise no real blockers.

File: pennylane/_qubit_device.py
Comment (josh146) [reply pennylane/_qubit_device.py:2023]: I'm curious if it should be the other way around? `adjoint_jacobian` checks if the new return type is enabled, and instead calls `adjoint_jacobian_new`/does post-processing?

BTW we should probably name it something more descriptive, like `adjoint_jacobian_tuple` 🙂

Comment (antalszava) [reply pennylane/_qubit_device.py:2023]: > I'm curious if it should be the other way around? adjoint_jacobian checks if the new return type is enabled, and instead calls adjoint_jacobian_new/does post-processing?

I think that works great too. :+1: :slightly_smiling_face: 

> BTW we should probably name it something more descriptive, like adjoint_jacobian_tuple slightly_smiling_face

Very fair point! We have some `_new` functions in the code base already, so likely best to come to a decision on naming for consistency. We'll be chatting about this in more depth via calls, saving this point for discussion and will return here.

Comment (josh146) [reply pennylane/_qubit_device.py:2023]: Ah got it! If it is doable, I think having `_tuple` might make more sense? But only for user/external-developer facing ones. Internal implementations I leave up to you :)

Comment (antalszava) [reply pennylane/_qubit_device.py]: It does :) It may just be that not all such functions return `_tuple` (for jacobians they should, however).

We've converged on a convention on code organization. Having said that, from my point of view, it's okay not to strictly follow it here because the changes required seem straightforward enough as they are now.

**Review by antalszava — APPROVED** [review]
Looks good :+1: 

Note: the exact implementation differs a bit from the convention we converged to. Having said that, I think it's okay to leave things as they are now.

File: tests/returntypes/test_adjoint_diff_new.py
Comment (rmoyard) [inline tests/returntypes/test_adjoint_diff_new.py:114]: Here we have to wait for finite diff to be merged.

**Review by rmoyard — APPROVED** [review]
Thanks @eddddddy for the changes in the jacobian method! It looks great 💯 I would just wait that both finite diff and param shift are merged before merging this one.

**Comment by rmoyard** [comment]
```
qml.enable_return()
dev = qml.device("default.qubit", wires=3)
params = np.array([np.pi, np.pi / 2, np.pi / 3])

with qml.tape.QuantumTape() as tape:
    qml.RX(params[0], wires=0)
    qml.RX(params[1], wires=1)
    qml.RX(params[2], wires=2)

    for idx in range(3):
        qml.expval(qml.PauliZ(idx))

# circuit jacobians
dev_jacobian = dev.adjoint_jacobian(tape)
print("Ajoint", dev_jacobian)

tapes, fn = qml.gradients.finite_diff(tape)
print("Finite diff", fn(dev.batch_execute(tapes)))

tapes, fn = qml.gradients.param_shift(tape)
print("Param shift", fn(dev.batch_execute(tapes)))
```
`Ajoint (array([-1.22464680e-16,  0.00000000e+00,  2.77555756e-17]), array([ 1.27035909e-33, -1.00000000e+00,  2.77555756e-17]), array([-4.08147434e-33,  5.55111512e-17, -8.66025404e-01]))`

`Finite diff ((array(5.10702591e-08), array(0.), array(-2.22044605e-09)), (array(0.), array(-1.), array(1.11022302e-09)), (array(-1.11022302e-09), array(1.11022302e-09), array(-0.86602543)))`

`Param shift ((array(-2.77555756e-16), array(0.), array(2.22044605e-16)), (array(0.), array(-1.), array(0.)), (array(0.), array(0.), array(-0.8660254)))`

@eddddddy @antalszava From the example below, I notice that there is a small problem with multiple measurement and multiple parameters. The shape of the inner structure should be multiple tuple when they are multiple measurements. Some more post-processing is needed to change that behavior.

**Review by rmoyard — CHANGES_REQUESTED** [review]
From the previous comment some changes are required to match the shape of the other custom gradient.

File: pennylane/_qubit_device.py
Comment (rmoyard) [inline pennylane/_qubit_device.py:1909]: In the case of numpy array I think you need to recall it on j because currently if is returning numpy float. I had the same issue with the finite difference case.

```
qml.enable_return()
import pennylane as qml

dev = qml.device("default.qubit", wires=2)

with qml.tape.QuantumTape() as tape:
    qml.RX(0.5, wires=0)
    qml.CNOT(wires=[0, 1])
    qml.RY(0.8, wires=1)
    qml.expval(qml.PauliZ(0))
    qml.expval(qml.PauliZ(1))
    qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))

jac = dev.adjoint_jacobian(tape)
```
`((-0.479425538604203, -3.469446951953614e-18), (-0.3340189893779267, -0.6295391960392662), (0.0, -0.7173560908995227))`

**Review by rmoyard — COMMENTED** [review]
Thanks @eddddddy for the changes! 💯 There is just a small convert issue for numpy array. Also I think it could be great to add (or find) some interfaces tests.

**Comment by rmoyard** [comment]
@eddddddy I think there might be some problems that come from the finite diff PR merge, let me know if you want me to take a look 👍

File: pennylane/_qubit_device.py
Comment (eddddddy) [reply pennylane/_qubit_device.py]: Done 👍

**Review by rmoyard — APPROVED** [review]
Thanks @eddddddy, great work it looks all good 💯

**Review by antalszava — APPROVED** [review]
:clap:

**Comment by rmoyard** [comment]
[sc-28430]

**Pull request merged**
